File: opus_edit.py
#!/usr/bin/env python3

import os
import shutil
import sys
import glob
import subprocess

import taglib

def progress_bar(current_pos, end_val, line=""):
    bar_length = 20
    full_size = bar_length + 7

    percent = current_pos / end_val
    hashes = "#" * round(percent * bar_length)
    spaces = " " * (bar_length - len(hashes))

    width = shutil.get_terminal_size()[0]
    line += " " * (width - full_size - len(line))
    status = "\r[{0}{1}]{2}% {3}".format(hashes, spaces,
                                  round(percent*100), line)

    if len(status) < width:
        print(status,end="")
    else:
        print(status[:width],end="")

    if current_pos == end_val:
        print()
        
def edit_tags(files, path):
    number_of_files = len(files)
    extension_length = len(".opus")
    current = 0

    for f in files:
        file_name = f[len(path):-extension_length]
        song = taglib.File(f)
        song.tags = {}

        artist, sep, title  = file_name.partition(" - ")
        song.tags["ARTIST"] = [artist]
        song.tags["TITLE"] = [title]
        song.save()

        current += 1
        progress_bar(current, number_of_files, file_name)

def refresh_mpd(pid):
    if os.path.exists(pid):
        print("Refreshing MPD")
        subprocess.call(["mpc","-q","clear"])
        subprocess.call(["mpc","-q","update"])
        subprocess.call(["mpc","-q","add","/"])
        subprocess.call(["mpc","-q","random","on"])
        subprocess.call(["mpc","-q","repeat","on"])
    else:
        print("MPD not currently running")

def move_files(files):
    try:
        dst = os.environ["HOME"] + "/Music/"
        [shutil.move(f,dst) for f in files]
        print("Moving files to ~/Music")
    except shutil.Error:
        print("Files already in Music directory")

if __name__ == "__main__":
    try:
        music_path = sys.argv[1]
        if not music_path.endswith("/"):
            music_path += "/"
    except IndexError:
        print("No path specified")
        exit()

    music_files = glob.glob(music_path + "*.opus")
    if music_files:
        edit_tags(music_files, music_path)
        print("Edited {0} songs".format(len(music_files)))

        move_files(music_files)
        mpd_pid = os.environ["HOME"] + "/.config/mpd/pid"
        refresh_mpd(mpd_pid)

        print("Done!")
    else:
        print("No files selected")

File: opus_edit.pl
#!/usr/bin/env perl

use strict;
use warnings;

use File::Copy;
use Audio::TagLib;

die "No path specified\n" unless @ARGV;
my @files = glob "$ARGV[0]/*.opus";

my $num = 0;
my $skip = 0;
my $curr = "";
my ($padd, $diff, $prev);

my ($f, $artist, $title, $prev_artist, $prev_title);

$| = 1;

for (@files) {
    $prev = $curr;
    ($curr) = /$ARGV[0]\/(.*)/;
    
    $diff = length($prev) - length($curr);
    $diff = 0 if $diff < 0;
    $padd = " " x $diff;

    print "\r[", ++$num, "/", $#files + 1, "] ",$curr, $padd;

    ($artist) = $curr =~ /(.*) -/; $artist =~ s/^\s+|\s+$//;
    ($title) = $curr =~ /- (.*)/; $title =~ s/^\s+|\.opus$//;

    $f = Audio::TagLib::FileRef->new($_);
    $prev_artist = $f->tag()->artist()->toCString();
    $prev_title = $f->tag()->title()->toCString();

    ++$skip and next if ($prev_artist eq $artist and $prev_title eq $title);

    $f->tag()->setArtist(Audio::TagLib::String->new($artist));
    $f->tag()->setTitle(Audio::TagLib::String->new($title));
    $f->save;
}

$| = 0;

if ($#files < 0) {
    die "No files selected\n";
} else {
    print "\nEdited ", $#files + 1 - $skip, " files($skip skipped)\n";
}

unless ("$ARGV[0]/" =~ /Music\//) {
    print "Moving files to music directory\n";
    move($_, "$ENV{HOME}/Music/") for @files;
} else {
    print "Files already in music directory\n";
}

if ($#files + 1 != $skip and -f "$ENV{HOME}/.config/mpd/pid") {
    print "Refreshing MPD\n";
    system "mpc -q clear && mpc -q update &&
            mpc -q add / && mpc -q random on &&
            mpc -q repeat on";
} else {
    print "Refreshing MPD not required\n";
}
